Enumeration type used for `std::integer_sequence` creation is undefined behaviour

```console
====================[ Build | SimpleExample | Debug ]===========================
"/Users/georgiy.lebedev/Library/Application Support/JetBrains/Toolbox/apps/CLion/ch-2/231.8109.222/CLion.app/Contents/bin/cmake/mac/bin/cmake" --build /Users/georgiy.lebedev/Work/tntcxx/build-debug --target SimpleExample -j 6
[1/2] Building CXX object CMakeFiles/SimpleExample.dir/examples/Simple.cpp.o
FAILED: CMakeFiles/SimpleExample.dir/examples/Simple.cpp.o 
ccache /usr/bin/clang++  -I/Users/georgiy.lebedev/Work/tntcxx/. -I/Users/georgiy.lebedev/Work/tntcxx/./src -I/Users/georgiy.lebedev/Work/tntcxx/./third_party/libev -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -fcolor-diagnostics -Wall -Wextra -Werror -std=gnu++17 -MD -MT CMakeFiles/SimpleExample.dir/examples/Simple.cpp.o -MF CMakeFiles/SimpleExample.dir/examples/Simple.cpp.o.d -o CMakeFiles/SimpleExample.dir/examples/Simple.cpp.o -c /Users/georgiy.lebedev/Work/tntcxx/examples/Simple.cpp
In file included from /Users/georgiy.lebedev/Work/tntcxx/examples/Simple.cpp:41:
In file included from /Users/georgiy.lebedev/Work/tntcxx/examples/../src/Client/Connector.hpp:32:
In file included from /Users/georgiy.lebedev/Work/tntcxx/examples/../src/Client/Connection.hpp:33:
In file included from /Users/georgiy.lebedev/Work/tntcxx/examples/../src/Client/RequestEncoder.hpp:32:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/any:92:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/memory:853:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__memory/compressed_pair.h:16:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/tuple:213:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__utility/integer_sequence.h:27:5: error: static_assert failed due to requirement 'is_integral<mpp::compact::Family>::value' "std::integer_sequence can only be instantiated with an integral type"
    static_assert( is_integral<_Tp>::value,
    ^              ~~~~~~~~~~~~~~~~~~~~~~~
/Users/georgiy.lebedev/Work/tntcxx/examples/../src/Client/../mpp/Dec.hpp:62:10: note: in instantiation of template class 'std::integer_sequence<mpp::compact::Family, mpp::compact::MP_NIL>' requested here
                return family_sequence<compact::MP_NIL>{};
```

https://github.com/tarantool/tntcxx/blob/0ce3f3292c842c2ad6930b7bb81e6add425115e8/src/mpp/Dec.hpp#L51-L52

According to the standard, the type used for `std::integer_sequence` is supposed to be an integer type[^1].

[`std::underlying_type`](https://en.cppreference.com/w/cpp/types/underlying_type) should be used to retrieve the enumeration type:

> The member typedef type names the underlying type of T.
Requires: T shall be a complete enumeration type[^2]

According to a GCC maintainer, this is considered undefined behaviour:
> The C++ standard requires that the first template argument for std::integer_sequence is an integer type, so your code is undefined.[^3] 



[^1]: https://timsong-cpp.github.io/cppwp/n4659/intseq.intseq
[^2]: https://timsong-cpp.github.io/cppwp/n4659/meta.trans.other
[^3]: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=88675

--- a/src/Client/Connection.hpp
+++ b/src/Client/Connection.hpp
@@ -46,6 +46,11 @@ static constexpr size_t CONN_READAHEAD = 64 * 1024;
 static constexpr size_t IOVEC_MAX_SIZE = 32;
 
 struct ConnectionError {
+	ConnectionError(const std::string &msg, int errno_ = 0) :
+		msg(msg), saved_errno(errno_)
+	{
+	}
+
 	std::string msg;
 	//Saved in case connection fails due to system error.
 	int saved_errno = 0;
@@ -87,7 +92,7 @@ public:
 	//It is useful to store connection objects in stl containers for example.
 	ssize_t refs;
 	//Members below can be default-initialized.
-	ConnectionError error;
+	std::optional<ConnectionError> error;
 	Greeting greeting;
 	bool is_greeting_received;
 	bool is_auth_required;
@@ -173,6 +178,7 @@ public:
 	rid_t ping();
 
 	void setError(const std::string &msg, int errno_ = 0);
+	bool hasError() const;
 	ConnectionError& getError();
 	void reset();
 	BUFFER& getInBuf();
@@ -408,22 +414,29 @@ template<class BUFFER, class NetProvider>
 void
 Connection<BUFFER, NetProvider>::setError(const std::string &msg, int errno_)
 {
-	impl->error.msg = msg;
-	impl->error.saved_errno = errno_;
+	impl->error.emplace(msg, errno_);
+}
+
+template<class BUFFER, class NetProvider>
+bool
+Connection<BUFFER, NetProvider>::hasError() const
+{
+	return impl->error.has_value();
 }
 
 template<class BUFFER, class NetProvider>
 ConnectionError&
 Connection<BUFFER, NetProvider>::getError()
 {
-	return impl->error;
+	assert(hasError());
+	return impl->error.value();
 }
 
 template<class BUFFER, class NetProvider>
 void
 Connection<BUFFER, NetProvider>::reset()
 {
-	impl->error = ConnectionError{};
+	impl->error.reset();
 }
 
 template<class BUFFER, class NetProvider>

--- a/src/Client/Connector.hpp
+++ b/src/Client/Connector.hpp
@@ -185,7 +185,8 @@ Connector<BUFFER, NetProvider>::wait(Connection<BUFFER, NetProvider> &conn,
 	timer.start();
 	if (connectionDecodeResponses(conn, result) != 0)
 		return -1;
-	while (! conn.futureIsReady(future) && !timer.isExpired()) {
+	while (!conn.hasError() && !conn.futureIsReady(future) &&
+	       !timer.isExpired()) {
 		if (m_NetProvider.wait(timeout - timer.elapsed()) != 0) {
 			conn.setError("Failed to poll: " + std::to_string(errno));
 			return -1;
@@ -201,6 +202,10 @@ Connector<BUFFER, NetProvider>::wait(Connection<BUFFER, NetProvider> &conn,
 			if (!hasDataToDecode(conn))
 				m_ReadyToDecode.erase(conn);
 		}
+	}
+	if (conn.hasError()) {
+		LOG_ERROR("Connection got an error: ", conn.getError().msg);
+		return -1;
 	}
 	if (! conn.futureIsReady(future)) {
 		LOG_ERROR("Connection has been timed out: future ", future,
@@ -220,7 +225,7 @@ Connector<BUFFER, NetProvider>::waitAll(Connection<BUFFER, NetProvider> &conn,
 	Timer timer{timeout};
 	timer.start();
 	size_t last_not_ready = 0;
-	while (!timer.isExpired()) {
+	while (!conn.hasError() && !timer.isExpired()) {
 		if (m_NetProvider.wait(timeout - timer.elapsed()) != 0) {
 			conn.setError("Failed to poll: " + std::to_string(errno));
 			return -1;
@@ -242,6 +247,10 @@ Connector<BUFFER, NetProvider>::waitAll(Connection<BUFFER, NetProvider> &conn,
 		}
 		if (finish)
 			return 0;
+	}
+	if (conn.hasError()) {
+		LOG_ERROR("Connection got an error: ", conn.getError().msg);
+		return -1;
 	}
 	LOG_ERROR("Connection has been timed out: not all futures are ready");
 	return -1;
@@ -276,7 +285,7 @@ Connector<BUFFER, NetProvider>::waitCount(Connection<BUFFER, NetProvider> &conn,
 	Timer timer{timeout};
 	timer.start();
 	size_t ready_futures = conn.getFutureCount();
-	while (!timer.isExpired()) {
+	while (!conn.hasError() && !timer.isExpired()) {
 		if (m_NetProvider.wait(timeout - timer.elapsed()) != 0) {
 			conn.setError("Failed to poll: " + std::to_string(errno));
 			return -1;
@@ -290,6 +299,10 @@ Connector<BUFFER, NetProvider>::waitCount(Connection<BUFFER, NetProvider> &conn,
 		}
 		if ((conn.getFutureCount() - ready_futures) >= future_count)
 			return 0;
+	}
+	if (conn.hasError()) {
+		LOG_ERROR("Connection got an error: ", conn.getError().msg);
+		return -1;
 	}
 	LOG_ERROR("Connection has been timed out: only ",
 		   conn.getFutureCount() - ready_futures, " are ready");

--- a/src/Client/EpollNetProvider.hpp
+++ b/src/Client/EpollNetProvider.hpp
@@ -232,11 +232,15 @@ EpollNetProvider<BUFFER, Stream>::send(Conn_t &conn)
 					    iov, IOVEC_MAX_SIZE);
 
 		ssize_t sent = conn.get_strm().send(iov, iov_cnt);
-		hasSentBytes(conn, (sent < 0 ? 0 : sent));
 		if (sent < 0) {
 			conn.setError(std::string("Failed to send request: ") +
-				      strerror(errno), errno);
+				      strerror(errno));
 			return -1;
+		} else if (sent == 0) {
+			assert(conn.get_strm().has_status(SS_NEED_WRITE_EVENT_FOR_WRITE));
+			return 1;
+		} else {
+			hasSentBytes(conn, sent);
 		}
 	}
 	/* All data from connection has been successfully written. */

--- a/src/Client/LibevNetProvider.hpp
+++ b/src/Client/LibevNetProvider.hpp
@@ -195,11 +195,15 @@ connectionSend(Connection<BUFFER,  LibevNetProvider<BUFFER, Stream>> &conn)
 					    iov, IOVEC_MAX_SIZE);
 
 		ssize_t sent = conn.get_strm().send(iov, iov_cnt);
-		hasSentBytes(conn, (sent < 0 ? 0 : sent));
 		if (sent < 0) {
 			conn.setError(std::string("Failed to send request: ") +
 				      strerror(errno));
 			return -1;
+		} else if (sent == 0) {
+			assert(conn.get_strm().has_status(SS_NEED_WRITE_EVENT_FOR_WRITE));
+			return 1;
+		} else {
+			hasSentBytes(conn, sent);
 		}
 	}
 	/* All data from connection has been successfully written. */

--- a/src/Client/UnixPlainStream.hpp
+++ b/src/Client/UnixPlainStream.hpp
@@ -106,7 +106,7 @@ UnixPlainStream::send(struct iovec *iov, size_t iov_count)
 	remove_status(SS_NEED_EVENT_FOR_WRITE);
 	struct msghdr msg = internal::create_msghdr(iov, iov_count);
 	while (true) {
-		int flags = MSG_DONTWAIT;
+		int flags = MSG_DONTWAIT | MSG_NOSIGNAL;
 		ssize_t sent = sendmsg(get_fd(), &msg, flags);
 
 		if (sent > 0)

--- a/src/Client/UnixSSLStream.hpp
+++ b/src/Client/UnixSSLStream.hpp
@@ -99,7 +99,7 @@ private:
 class UnixSSLStream : public UnixPlainStream {
 public:
 	UnixSSLStream() noexcept = default;
-	inline ~UnixSSLStream() noexcept;
+	inline ~UnixSSLStream();
 	UnixSSLStream(const UnixSSLStream&) = delete;
 	UnixSSLStream &operator=(const UnixSSLStream&) = delete;
 	UnixSSLStream(UnixSSLStream &&a) noexcept = default;

--- a/src/Client/UnixSSLStream.hpp
+++ b/src/Client/UnixSSLStream.hpp
@@ -354,6 +354,12 @@ int UnixSSLStream::connect(const ConnectOptions &opts_arg)
 		return 0;
 	assert(opts.transport == STREAM_SSL);
 
+	int opt = 1;
+	if (::setsockopt(get_fd(), SOL_SOCKET, SO_NOSIGPIPE,
+			 &opt, sizeof(opt)) != 0) {
+		return US_DIE("setsockopt failed", strerror(errno));
+	}
+
 	if (ssl_context.create(opts) != 0)
 		return US_DIE("SSL_context create failed",
 			      ssl_context.get_last_error());

--- a/src/Utils/List.hpp
+++ b/src/Utils/List.hpp
@@ -204,9 +204,14 @@ private:
 	// iterator and const_iterator below.
 	template <class IElem, class IRing>
 	class iterator_common
-		: std::iterator<std::bidirectional_iterator_tag, IElem>
 	{
 	public:
+		using iterator_category = std::bidirectional_iterator_tag;
+		using value_type = IElem;
+		using difference_type = std::ptrdiff_t;
+		using pointer = IElem *;
+		using reference = IElem &;
+
 		IElem& operator*() const noexcept;
 		IElem *operator->() const noexcept;
 		bool operator==(const iterator_common& aItr) const noexcept;

--- a/src/mpp/Dec.hpp
+++ b/src/mpp/Dec.hpp
@@ -49,7 +49,7 @@ namespace literal = tnt::literal;
 namespace decode_details {
 
 template <compact::Family ...sequence>
-using family_sequence = std::integer_sequence<compact::Family, sequence...>;
+using family_sequence = std::integer_sequence<std::underlying_type_t<compact::Family>, sequence...>;
 
 template <class T>
 constexpr auto detectFamily()

--- a/test/ClientTest.cpp
+++ b/test/ClientTest.cpp
@@ -37,6 +37,7 @@
 
 const char *localhost = "127.0.0.1";
 int port = 3301;
+int dummy_server_port = 3302;
 const char *unixsocket = "./tnt.sock";
 int WAIT_TIMEOUT = 1000; //milliseconds
 
@@ -693,6 +694,30 @@ test_auth(Connector<BUFFER, NetProvider> &client)
 	printResponse<BUFFER, NetProvider>(conn, *response);
 }
 
+/** Single connection, write to closed connection. */
+template <class BUFFER, class NetProvider>
+void
+test_sigpipe(Connector<BUFFER, NetProvider> &client)
+{
+	TEST_INIT(0);
+
+	int rc = ::launchDummyServer(localhost, dummy_server_port);
+	fail_unless(rc == 0);
+
+	Connection<Buf_t, NetProvider> conn(client);
+	rc = ::test_connect(client, conn, localhost, dummy_server_port);
+	fail_unless(rc == 0);
+
+	/*
+	 * Create a large payload so that request needs at least 2 `send`s, the
+	 * latter being written to a closed socket.
+	 */
+	rid_t f = conn.space[0].replace(std::vector<uint64_t>(100000, 777));
+	fail_if(client.wait(conn, f, WAIT_TIMEOUT) == 0);
+	fail_unless(conn.getError().saved_errno == EPIPE);
+	fail_if(conn.futureIsReady(f));
+}
+
 int main()
 {
 	if (cleanDir() != 0)
@@ -723,5 +748,12 @@ int main()
 	single_conn_call<Buf_t, NetProvider>(client);
 	replace_unix_socket(client);
 	test_auth(client);
+	/*
+	 * Testing this for SSL is hard, since the connection starts to involve
+	 * an a lot more complex state machine.
+	 */
+#ifndef TNTCXX_ENABLE_SSL
+	::test_sigpipe(client);
+#endif
 	return 0;
 }

--- a/test/ClientTest.cpp
+++ b/test/ClientTest.cpp
@@ -48,6 +48,12 @@ constexpr bool enable_ssl = false;
 constexpr StreamTransport transport = STREAM_PLAIN;
 #endif
 
+#ifdef __linux__
+using NetProvider = EpollNetProvider<Buf_t, DefaultStream>;
+#else
+using NetProvider = LibevNetProvider<Buf_t, DefaultStream>;
+#endif
+
 template <class Connector, class Connection>
 static int
 test_connect(Connector &client, Connection &conn, const std::string &addr,
@@ -701,40 +707,21 @@ int main()
 		return -1;
 
 	sleep(1);
-#ifdef __linux__
-	using NetEpoll_t = EpollNetProvider<Buf_t, DefaultStream>;
-	Connector<Buf_t, NetEpoll_t> client;
-	trivial<Buf_t, NetEpoll_t>(client);
-	single_conn_ping<Buf_t, NetEpoll_t>(client);
-	auto_close<Buf_t, NetEpoll_t>(client);
-	many_conn_ping<Buf_t, NetEpoll_t>(client);
-	single_conn_error<Buf_t, NetEpoll_t>(client);
-	single_conn_replace<Buf_t, NetEpoll_t>(client);
-	single_conn_insert<Buf_t, NetEpoll_t>(client);
-	single_conn_update<Buf_t, NetEpoll_t>(client);
-	single_conn_delete<Buf_t, NetEpoll_t>(client);
-	single_conn_upsert<Buf_t, NetEpoll_t>(client);
-	single_conn_select<Buf_t, NetEpoll_t>(client);
-	single_conn_call<Buf_t, NetEpoll_t>(client);
+
+	Connector<Buf_t, NetProvider> client;
+	trivial<Buf_t, NetProvider>(client);
+	single_conn_ping<Buf_t, NetProvider>(client);
+	auto_close<Buf_t, NetProvider>(client);
+	many_conn_ping<Buf_t, NetProvider>(client);
+	single_conn_error<Buf_t, NetProvider>(client);
+	single_conn_replace<Buf_t, NetProvider>(client);
+	single_conn_insert<Buf_t, NetProvider>(client);
+	single_conn_update<Buf_t, NetProvider>(client);
+	single_conn_delete<Buf_t, NetProvider>(client);
+	single_conn_upsert<Buf_t, NetProvider>(client);
+	single_conn_select<Buf_t, NetProvider>(client);
+	single_conn_call<Buf_t, NetProvider>(client);
 	replace_unix_socket(client);
 	test_auth(client);
-#endif
-	/* LibEv network provide */
-	using NetLibEv_t = LibevNetProvider<Buf_t, DefaultStream>;
-	Connector<Buf_t, NetLibEv_t > another_client;
-	trivial<Buf_t, NetLibEv_t >(another_client);
-	single_conn_ping<Buf_t, NetLibEv_t>(another_client);
-	auto_close<Buf_t, NetLibEv_t>(another_client);
-	many_conn_ping<Buf_t, NetLibEv_t>(another_client);
-	single_conn_error<Buf_t, NetLibEv_t>(another_client);
-	single_conn_replace<Buf_t, NetLibEv_t>(another_client);
-	single_conn_insert<Buf_t, NetLibEv_t>(another_client);
-	single_conn_update<Buf_t, NetLibEv_t>(another_client);
-	single_conn_delete<Buf_t, NetLibEv_t>(another_client);
-	single_conn_upsert<Buf_t, NetLibEv_t>(another_client);
-	single_conn_select<Buf_t, NetLibEv_t>(another_client);
-	single_conn_call<Buf_t, NetLibEv_t>(another_client);
-	replace_unix_socket(another_client);
-	test_auth(another_client);
 	return 0;
 }

--- a/test/ClientTest.cpp
+++ b/test/ClientTest.cpp
@@ -718,6 +718,37 @@ test_sigpipe(Connector<BUFFER, NetProvider> &client)
 	fail_if(conn.futureIsReady(f));
 }
 
+/** Single connection, wait response from closed connection. */
+template <class BUFFER, class NetProvider>
+void
+test_dead_connection_wait(Connector<BUFFER, NetProvider> &client)
+{
+	TEST_INIT(0);
+
+	int rc = ::launchDummyServer(localhost, dummy_server_port);
+	fail_unless(rc == 0);
+
+	Connection<Buf_t, NetProvider> conn(client);
+	rc = ::test_connect(client, conn, localhost, dummy_server_port);
+	fail_unless(rc == 0);
+
+	rid_t f = conn.ping();
+	fail_if(client.wait(conn, f) == 0);
+	fail_if(conn.futureIsReady(f));
+
+	fail_if(client.waitAll(conn, std::vector<rid_t>(f)) == 0);
+	fail_if(conn.futureIsReady(f));
+
+	fail_if(client.waitCount(conn, 1) == 0);
+	fail_if(conn.futureIsReady(f));
+
+	/* FIXME(gh-51) */
+#if 0
+	fail_if(client.waitAny() != std::nullopt);
+	fail_if(conn.futureIsReady(f));
+#endif
+}
+
 int main()
 {
 	if (cleanDir() != 0)
@@ -755,5 +786,6 @@ int main()
 #ifndef TNTCXX_ENABLE_SSL
 	::test_sigpipe(client);
 #endif
+	::test_dead_connection_wait(client);
 	return 0;
 }

--- a/test/Utils/System.hpp
+++ b/test/Utils/System.hpp
@@ -29,10 +29,13 @@
  * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
+#include <arpa/inet.h>
+#include <netinet/in.h>
 #include <unistd.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <sys/wait.h>
+#include <sys/socket.h>
 #include <signal.h>
 #ifdef __linux__
 #include <sys/prctl.h>
@@ -128,4 +131,82 @@ genSSLCert() {
 			strerror(errno));
 	}
 	exit(EXIT_FAILURE);
+}
+
+/**
+ * Fork a dummy server that accepts one connection on
+ * `localhost:dummy_server_port`, reads 1 byte and exits.
+ */
+int
+launchDummyServer(const char *addr, int port)
+{
+	pid_t ppid_before_fork = ::getpid();
+	pid_t pid = ::fork();
+	if (pid < 0) {
+		::perror("fork failed");
+		return -1;
+	} else if (pid != 0) {
+		::sleep(1);
+		return 0;
+	}
+
+	set_parent_death_signal(ppid_before_fork, "dummy server");
+
+	int sock = ::socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	if (sock < 0) {
+		::perror("socket failed");
+		::exit(EXIT_FAILURE);
+	}
+
+	int opt = 1;
+	if (::setsockopt(sock, SOL_SOCKET, SO_REUSEADDR,
+		       &opt, sizeof(opt)) != 0) {
+		::perror("setsockopt failed");
+		::exit(EXIT_FAILURE);
+	}
+	if (::setsockopt(sock, SOL_SOCKET, SO_REUSEPORT,
+		       &opt, sizeof(opt)) != 0) {
+		::perror("setsockopt failed");
+		::exit(EXIT_FAILURE);
+	}
+
+	struct sockaddr_in sock_address = {
+		.sin_family = AF_INET,
+		.sin_port = htons(port),
+	};
+	if (::inet_aton(addr, &sock_address.sin_addr) != 1) {
+		::perror("inet_aton failed");
+		::exit(EXIT_FAILURE);
+	}
+
+	if (::bind(sock,
+		   reinterpret_cast<const struct sockaddr *>(&sock_address),
+		   sizeof(sockaddr)) != 0) {
+		::perror("bind failed");
+		::exit(EXIT_FAILURE);
+	}
+
+	if (::listen(sock, 1) != 0) {
+		::perror("listen failed");
+		::exit(EXIT_FAILURE);
+	}
+
+	int accepted_sock = ::accept(sock, nullptr, nullptr);
+	if (accepted_sock < 0) {
+		::perror("accept failed");
+		::exit(EXIT_FAILURE);
+	}
+
+	::close(sock);
+
+	/* Allow first `send` call to this socket to succeed. */
+	char buf[1];
+	if (::read(accepted_sock, buf, sizeof(buf)) < 0) {
+		::perror("read failed");
+		::exit(EXIT_FAILURE);
+	}
+
+	::close(accepted_sock);
+
+	::exit(EXIT_SUCCESS);
 }

--- a/test/Utils/System.hpp
+++ b/test/Utils/System.hpp
@@ -38,6 +38,27 @@
 #include <sys/prctl.h>
 #endif
 
+void
+set_parent_death_signal(pid_t ppid_before_fork, const char *child_program_name)
+{
+	/**
+	 * Kill the child (the current process) when the test (the parent
+	 * process) is finished.
+	 */
+#ifdef __linux__
+	if (prctl(PR_SET_PDEATHSIG, SIGTERM) == -1) {
+		fprintf(stderr, "Can't launch %s: %s", child_program_name,
+			strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+#endif
+	if (getppid() != ppid_before_fork) {
+		fprintf(stderr, "Can't launch %s: parent process exited "
+				"just before prctl call", child_program_name);
+		exit(EXIT_FAILURE);
+	}
+}
+
 inline int
 launchTarantool(bool enable_ssl = false)
 {
@@ -51,21 +72,7 @@ launchTarantool(bool enable_ssl = false)
 	/* Returning from parent. */
 	if (pid != 0)
 		return 0;
-	/*
-	 * Kill child (i.e. Tarantool process) when the test is finished.
-	 * Doesn't work on MacOS tho.
-	 */
-#ifdef __linux__
-	if (prctl(PR_SET_PDEATHSIG, SIGTERM) == -1) {
-		fprintf(stderr, "Can't launch Tarantool: %s", strerror(errno));
-		exit(EXIT_FAILURE);
-	}
-#endif
-	if (getppid() != ppid_before_fork) {
-		fprintf(stderr, "Can't launch Tarantool: parent process exited "\
-				"just before prctl call");
-		exit(EXIT_FAILURE);
-	}
+	set_parent_death_signal(ppid_before_fork, "Tarantool");
 	const char *script = enable_ssl ? "test_cfg_ssl.lua" : "test_cfg.lua";
 	if (execlp("tarantool", "tarantool", script, NULL) == -1) {
 		fprintf(stderr, "Can't launch Tarantool: execlp failed! %s\n",

